compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1427,46 +1427,207 @@ public boolean isAssignableToFiniteType(BType type,
+                        if (rhsMemberType.tag == TypeTags.STRING || rhsMemberType.tag == TypeTags.INT ||
+                                rhsMemberType.tag == TypeTags.FLOAT || rhsMemberType.tag == TypeTags.BOOLEAN ||
+                                rhsMemberType.tag == TypeTags.NIL) {
+                            matchFound = true;
+                            break;
+                        }
+                    }
+                    break;
+                // When expanding members for tuples, arrays and maps, set isValueDeepEquality to true, to allow
+                // comparison between JSON lists/maps and primitive lists/maps since they are all reference types
+                case TypeTags.TUPLE:
+                    for (BType rhsMemberType : rhsTypes) {
+                        if (rhsMemberType.tag == TypeTags.TUPLE &&
+                                tupleIntersectionExists((BTupleType) lhsMemberType, (BTupleType) rhsMemberType,
+                                                        true)) {
+                            matchFound = true;
+                            break;
+                        }
+                    }
+                    break;
+                case TypeTags.ARRAY:
+                    for (BType rhsMemberType : rhsTypes) {
+                        if (rhsMemberType.tag == TypeTags.ARRAY &&
+                                intersectionExists(expandAndGetMemberTypesRecursive(((BArrayType) lhsMemberType).eType),
+                                                   expandAndGetMemberTypesRecursive(((BArrayType) rhsMemberType).eType),
+                                                   true)) {
+                            matchFound = true;
+                            break;
+                        }
+                    }
+                    break;
+                case TypeTags.MAP:
+                    for (BType rhsMemberType : rhsTypes) {
+                        if (rhsMemberType.tag == TypeTags.MAP && intersectionExists(
+                                expandAndGetMemberTypesRecursive(((BMapType) lhsMemberType).constraint),
+                                expandAndGetMemberTypesRecursive(((BMapType) rhsMemberType).constraint), true)) {
+                            matchFound = true;
+                            break;
+                        }
+                    }
+                    break;
+                case TypeTags.OBJECT:
+                case TypeTags.RECORD:
+                    for (BType rhsMemberType : rhsTypes) {
+                        if (checkStructEquivalency(rhsMemberType, lhsMemberType) ||
+                                checkStructEquivalency(lhsMemberType, rhsMemberType)) {
+                            // intersection exists if one is assignable to the other
+                            matchFound = true;
+                            break;
+                        }
+                    }
+                    break;
```

Comment:
we can stream and do  anyMatch(). It will simplify the code.

Can return if a match is found, then and there.

Reply:
Refactored the method and addressed all suggestions, thanks!